File: packages/sdk-go/csat_test.go
package pocketping

import (
	"context"
	"errors"
	"strings"
	"sync"
	"testing"
	"time"
)

// notifyCall records a Notify invocation on the test bridge.
type notifyCall struct {
	session *Session
	message string
}

// NotifyBridge implements Bridge + BridgeWithNotify, capturing Notify calls.
type NotifyBridge struct {
	BaseBridge
	mu          sync.Mutex
	notifyCalls []notifyCall
	notifyErr   error
}

func newNotifyBridge() *NotifyBridge {
	return &NotifyBridge{BaseBridge: BaseBridge{BridgeName: "telegram"}}
}

func (b *NotifyBridge) Notify(ctx context.Context, session *Session, message string) error {
	b.mu.Lock()
	defer b.mu.Unlock()
	b.notifyCalls = append(b.notifyCalls, notifyCall{session: session, message: message})
	return b.notifyErr
}

func (b *NotifyBridge) lastNotify() (notifyCall, bool) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if len(b.notifyCalls) == 0 {
		return notifyCall{}, false
	}
	return b.notifyCalls[len(b.notifyCalls)-1], true
}

func newCsatSession(t *testing.T, pp *PocketPing) string {
	t.Helper()
	resp, err := pp.HandleConnect(context.Background(), ConnectRequest{VisitorID: "v1"})
	if err != nil {
		t.Fatalf("HandleConnect failed: %v", err)
	}
	return resp.SessionID
}

func TestRequestCsat_SetsPendingAndBroadcasts(t *testing.T) {
	ctx := context.Background()
	bridge := newNotifyBridge()
	pp := New(Config{Bridges: []Bridge{bridge}})
	sessionID := newCsatSession(t, pp)

	conn := &MockWebSocketConn{}
	pp.RegisterWebSocket(sessionID, conn)

	if err := pp.RequestCsat(ctx, sessionID); err != nil {
		t.Fatalf("RequestCsat failed: %v", err)
	}

	session, _ := pp.GetSession(ctx, sessionID)
	if session.Csat == nil || !session.Csat.Pending {
		t.Fatalf("expected csat.pending true, got %+v", session.Csat)
	}
	if session.Csat.RequestedAt == nil {
		t.Errorf("expected csat.requestedAt to be set")
	}

	msgs := conn.GetMessages()
	if len(msgs) != 1 {
		t.Fatalf("expected 1 broadcast, got %d", len(msgs))
	}
	event, ok := msgs[0].(WebSocketEvent)
	if !ok || event.Type != "csat_request" {
		t.Errorf("expected csat_request event, got %+v", msgs[0])
	}
}

func TestHandleCsat_StoresScoreClearsPendingNotifiesRunsCallback(t *testing.T) {
	ctx := context.Background()
	bridge := newNotifyBridge()

	var (
		callbackMu      sync.Mutex
		callbackSession *Session
		callbackRating  CsatRating
	)
	pp := New(Config{
		Bridges: []Bridge{bridge},
		OnCsat: func(session *Session, rating CsatRating) {
			callbackMu.Lock()
			defer callbackMu.Unlock()
			callbackSession = session
			callbackRating = rating
		},
	})

	sessionID := newCsatSession(t, pp)
	if err := pp.RequestCsat(ctx, sessionID); err != nil {
		t.Fatalf("RequestCsat failed: %v", err)
	}

	res, err := pp.HandleCsat(ctx, CsatRequest{SessionID: sessionID, Score: 5, Comment: "  great  "})
	if err != nil {
		t.Fatalf("HandleCsat failed: %v", err)
	}
	if !res.OK || res.AlreadyRated {
		t.Errorf("expected {OK:true}, got %+v", res)
	}

	session, _ := pp.GetSession(ctx, sessionID)
	if session.Csat == nil {
		t.Fatal("expected csat state")
	}
	if session.Csat.Score == nil || *session.Csat.Score != 5 {
		t.Errorf("expected score 5, got %v", session.Csat.Score)
	}
	if session.Csat.Comment == nil || *session.Csat.Comment != "great" {
		t.Errorf("expected trimmed comment 'great', got %v", session.Csat.Comment)
	}
	if session.Csat.Pending {
		t.Errorf("expected pending false")
	}
	if session.Csat.RespondedAt == nil {
		t.Errorf("expected respondedAt to be set")
	}

	last, ok := bridge.lastNotify()
	if !ok {
		t.Fatal("expected a bridge notify call")
	}
	want := `⭐ 😍 5/5 — "great"`
	if last.message != want {
		t.Errorf("expected notify message %q, got %q", want, last.message)
	}

	callbackMu.Lock()
	defer callbackMu.Unlock()
	if callbackSession == nil {
		t.Fatal("expected OnCsat callback to run")
	}
	if callbackRating.Score != 5 || callbackRating.Comment != "great" {
		t.Errorf("expected rating {5, great}, got %+v", callbackRating)
	}
}

func TestHandleCsat_RejectsOutOfRangeScore(t *testing.T) {
	ctx := context.Background()
	pp := New(Config{})
	sessionID := newCsatSession(t, pp)

	for _, score := range []int{0, 6} {
		_, err := pp.HandleCsat(ctx, CsatRequest{SessionID: sessionID, Score: score})
		if err == nil {
			t.Fatalf("expected error for score %d", score)
		}
		if !strings.Contains(err.Error(), "1-5") {
			t.Errorf("expected error to mention 1-5, got %v", err)
		}
	}
}

func TestHandleCsat_IdempotentOnceRated(t *testing.T) {
	ctx := context.Background()
	pp := New(Config{})
	sessionID := newCsatSession(t, pp)

	if _, err := pp.HandleCsat(ctx, CsatRequest{SessionID: sessionID, Score: 4}); err != nil {
		t.Fatalf("first HandleCsat failed: %v", err)
	}

	second, err := pp.HandleCsat(ctx, CsatRequest{SessionID: sessionID, Score: 1})
	if err != nil {
		t.Fatalf("second HandleCsat failed: %v", err)
	}
	if !second.OK || !second.AlreadyRated {
		t.Errorf("expected {OK:true, AlreadyRated:true}, got %+v", second)
	}

	session, _ := pp.GetSession(ctx, sessionID)
	if session.Csat.Score == nil || *session.Csat.Score != 4 {
		t.Errorf("expected score unchanged at 4, got %v", session.Csat.Score)
	}
}

func TestCsat_SessionNotFound(t *testing.T) {
	ctx := context.Background()
	pp := New(Config{})

	if _, err := pp.HandleCsat(ctx, CsatRequest{SessionID: "nope", Score: 3}); !errors.Is(err, ErrSessionNotFound) {
		t.Errorf("expected ErrSessionNotFound from HandleCsat, got %v", err)
	}
	if err := pp.RequestCsat(ctx, "nope"); !errors.Is(err, ErrSessionNotFound) {
		t.Errorf("expected ErrSessionNotFound from RequestCsat, got %v", err)
	}
}

func TestGetStats_ComputesOverStorage(t *testing.T) {
	ctx := context.Background()
	pp := New(Config{})

	a, err := pp.HandleConnect(ctx, ConnectRequest{VisitorID: "va"})
	if err != nil {
		t.Fatalf("connect a failed: %v", err)
	}
	b, err := pp.HandleConnect(ctx, ConnectRequest{VisitorID: "vb"})
	if err != nil {
		t.Fatalf("connect b failed: %v", err)
	}

	// Session A: visitor msg + operator reply + 5-star rating.
	if _, err := pp.HandleMessage(ctx, SendMessageRequest{SessionID: a.SessionID, Content: "hi", Sender: SenderVisitor}); err != nil {
		t.Fatalf("message a/visitor failed: %v", err)
	}
	if _, err := pp.SendOperatorMessage(ctx, a.SessionID, "hello!", "", ""); err != nil {
		t.Fatalf("operator message failed: %v", err)
	}
	if _, err := pp.HandleCsat(ctx, CsatRequest{SessionID: a.SessionID, Score: 5}); err != nil {
		t.Fatalf("csat a failed: %v", err)
	}

	// Session B: visitor msg only (unanswered).
	if _, err := pp.HandleMessage(ctx, SendMessageRequest{SessionID: b.SessionID, Content: "anyone?", Sender: SenderVisitor}); err != nil {
		t.Fatalf("message b/visitor failed: %v", err)
	}

	stats, err := pp.GetStats(ctx, nil)
	if err != nil {
		t.Fatalf("GetStats failed: %v", err)
	}

	if stats.Conversations != 2 {
		t.Errorf("expected 2 conversations, got %d", stats.Conversations)
	}
	if stats.ResponseRate != 0.5 {
		t.Errorf("expected response rate 0.5, got %v", stats.ResponseRate)
	}
	if stats.UnansweredNow != 1 {
		t.Errorf("expected 1 unanswered, got %d", stats.UnansweredNow)
	}
	if stats.Csat.Responses != 1 {
		t.Errorf("expected 1 csat response, got %d", stats.Csat.Responses)
	}
	if stats.Csat.Percent == nil || *stats.Csat.Percent != 1 {
		t.Errorf("expected csat percent 1, got %v", stats.Csat.Percent)
	}
	if stats.Csat.Average == nil || *stats.Csat.Average != 5 {
		t.Errorf("expected csat average 5, got %v", stats.Csat.Average)
	}
	if len(stats.ConversationsSparkline) != 7 {
		t.Errorf("expected sparkline length 7, got %d", len(stats.ConversationsSparkline))
	}
}

// statsLessStorage is a Storage that does NOT implement StorageWithListSessions.
type statsLessStorage struct{}

func (statsLessStorage) CreateSession(ctx context.Context, session *Session) error { return nil }
func (statsLessStorage) GetSession(ctx context.Context, sessionID string) (*Session, error) {
	return nil, nil
}
func (statsLessStorage) GetSessionByVisitorID(ctx context.Context, visitorID string) (*Session, error) {
	return nil, nil
}
func (statsLessStorage) UpdateSession(ctx context.Context, session *Session) error { return nil }
func (statsLessStorage) DeleteSession(ctx context.Context, sessionID string) error { return nil }
func (statsLessStorage) SaveMessage(ctx context.Context, message *Message) error   { return nil }
func (statsLessStorage) GetMessages(ctx context.Context, sessionID, after string, limit int) ([]Message, error) {
	return nil, nil
}
func (statsLessStorage) GetMessage(ctx context.Context, messageID string) (*Message, error) {
	return nil, nil
}
func (statsLessStorage) CleanupOldSessions(ctx context.Context, olderThan time.Time) (int, error) {
	return 0, nil
}

func TestGetStats_UnsupportedStorage(t *testing.T) {
	ctx := context.Background()
	pp := New(Config{Storage: statsLessStorage{}})

	_, err := pp.GetStats(ctx, nil)
	if err == nil {
		t.Fatal("expected error for storage without ListSessions")
	}
	if !strings.Contains(err.Error(), "listSessions") {
		t.Errorf("expected error to mention listSessions, got %v", err)
	}
}

File: packages/sdk-php/tests/CsatTest.php
<?php

declare(strict_types=1);

namespace PocketPing\Tests;

use PHPUnit\Framework\TestCase;
use PocketPing\Bridges\AbstractBridge;
use PocketPing\Models\ConnectRequest;
use PocketPing\Models\CsatRequest;
use PocketPing\Models\Sender;
use PocketPing\Models\SendMessageRequest;
use PocketPing\Models\Session;
use PocketPing\PocketPing;
use PocketPing\Storage\MemoryStorage;

/**
 * Bridge that records plain notifications (the CSAT one-liner channel).
 */
final class NotifyBridge extends AbstractBridge
{
    /** @var array<array{session: Session, message: string}> */
    public array $disconnectCalls = [];

    public function getName(): string
    {
        return 'telegram';
    }

    public function notifyDisconnect(Session $session, string $message): void
    {
        $this->disconnectCalls[] = ['session' => $session, 'message' => $message];
    }
}

/**
 * Websocket double that records every payload it is sent.
 */
final class RecordingSocket
{
    /** @var string[] */
    public array $sent = [];

    public function send(string $message): void
    {
        $this->sent[] = $message;
    }
}

class CsatTest extends TestCase
{
    private NotifyBridge $bridge;
    private PocketPing $pp;

    protected function setUp(): void
    {
        $this->bridge = new NotifyBridge();
        $this->pp = new PocketPing(bridges: [$this->bridge]);
    }

    private function newSession(): string
    {
        return $this->pp->handleConnect(new ConnectRequest(visitorId: 'v1'))->sessionId;
    }

    public function testRequestCsatSetsPendingAndBroadcastsCsatRequest(): void
    {
        $sessionId = $this->newSession();
        $ws = new RecordingSocket();
        $this->pp->registerWebsocket($sessionId, $ws);

        $result = $this->pp->requestCsat($sessionId);
        $this->assertSame(['ok' => true], $result);

        $session = $this->pp->getSession($sessionId);
        $this->assertNotNull($session?->csat);
        $this->assertTrue($session->csat->pending);
        $this->assertInstanceOf(\DateTimeImmutable::class, $session->csat->requestedAt);

        $this->assertNotEmpty($ws->sent);
        $this->assertStringContainsString('csat_request', $ws->sent[0]);
    }

    public function testHandleCsatStoresScoreClearsPendingNotifiesBridgeRunsOnCsat(): void
    {
        /** @var array{0: Session, 1: array{score: int, comment: ?string}}|null $captured */
        $captured = null;
        $this->pp = new PocketPing(
            bridges: [$this->bridge],
            onCsat: function (Session $session, array $rating) use (&$captured): void {
                $captured = [$session, $rating];
            },
        );

        $sessionId = $this->newSession();
        $this->pp->requestCsat($sessionId);

        $res = $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 5, comment: '  great  '));
        $this->assertSame(['ok' => true], $res->toArray());

        $session = $this->pp->getSession($sessionId);
        $this->assertSame(5, $session?->csat?->score);
        $this->assertSame('great', $session->csat->comment);
        $this->assertFalse($session->csat->pending);
        $this->assertInstanceOf(\DateTimeImmutable::class, $session->csat->respondedAt);

        $last = $this->bridge->disconnectCalls[count($this->bridge->disconnectCalls) - 1] ?? null;
        $this->assertSame('⭐ 😍 5/5 — "great"', $last['message'] ?? null);

        $this->assertNotNull($captured);
        $this->assertSame(['score' => 5, 'comment' => 'great'], $captured[1]);
    }

    public function testHandleCsatAcceptsArrayBody(): void
    {
        $sessionId = $this->newSession();

        $res = $this->pp->handleCsat(['sessionId' => $sessionId, 'score' => 3]);
        $this->assertSame(['ok' => true], $res->toArray());

        $session = $this->pp->getSession($sessionId);
        $this->assertSame(3, $session?->csat?->score);
        $this->assertNull($session->csat->comment);
    }

    public function testHandleCsatRejectsOutOfRangeScore(): void
    {
        $sessionId = $this->newSession();

        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessageMatches('/1-5/');
        $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 0));
    }

    public function testHandleCsatRejectsScoreAboveRange(): void
    {
        $sessionId = $this->newSession();

        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessageMatches('/1-5/');
        $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 6));
    }

    public function testHandleCsatRejectsFractionalScore(): void
    {
        $sessionId = $this->newSession();

        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessageMatches('/1-5/');
        // 3.9 must be rejected outright, not silently truncated to 3.
        $this->pp->handleCsat(['sessionId' => $sessionId, 'score' => 3.9]);
    }

    public function testHandleCsatRejectsNonNumericScore(): void
    {
        $sessionId = $this->newSession();

        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessageMatches('/1-5/');
        $this->pp->handleCsat(['sessionId' => $sessionId, 'score' => 'good']);
    }

    public function testHandleCsatIsIdempotentOnceRated(): void
    {
        $sessionId = $this->newSession();

        $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 4));
        $second = $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 1));

        $this->assertSame(['ok' => true, 'alreadyRated' => true], $second->toArray());

        $session = $this->pp->getSession($sessionId);
        $this->assertSame(4, $session?->csat?->score); // unchanged
    }

    public function testHandleCsatThrowsWhenSessionMissing(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage('Session not found');
        $this->pp->handleCsat(new CsatRequest(sessionId: 'nope', score: 3));
    }

    public function testRequestCsatThrowsWhenSessionMissing(): void
    {
        $this->expectException(\InvalidArgumentException::class);
        $this->expectExceptionMessage('Session not found');
        $this->pp->requestCsat('nope');
    }

    public function testHandleCsatFiresSignedWebhook(): void
    {
        $http = new MockHttpClient();
        $this->pp = new PocketPing(
            webhookUrl: 'https://example.com/hook',
            webhookSecret: 'topsecret',
            httpClient: $http,
        );

        $sessionId = $this->newSession();
        $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 5, comment: 'nice'));

        $req = $http->getLastRequest();
        $this->assertNotNull($req);
        $this->assertSame('POST', $req['method']);
        $this->assertSame('https://example.com/hook', $req['url']);
        $this->assertSame('csat_submitted', $req['data']['type']);
        $this->assertSame($sessionId, $req['data']['data']['sessionId']);
        $this->assertSame(5, $req['data']['data']['score']);
        $this->assertSame('nice', $req['data']['data']['comment']);
        $this->assertArrayHasKey('respondedAt', $req['data']['data']);
        $this->assertArrayHasKey('sentAt', $req['data']);

        // Signature is sha256=HMAC over the JSON-encoded payload.
        $this->assertArrayHasKey('X-PocketPing-Signature', $req['headers']);
        $body = json_encode($req['data'], JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);
        $expected = 'sha256=' . hash_hmac('sha256', (string) $body, 'topsecret');
        $this->assertSame($expected, $req['headers']['X-PocketPing-Signature']);
    }

    public function testHandleCsatSkipsWebhookWhenNoUrlConfigured(): void
    {
        $http = new MockHttpClient();
        $this->pp = new PocketPing(httpClient: $http);

        $sessionId = $this->newSession();
        $this->pp->handleCsat(new CsatRequest(sessionId: $sessionId, score: 4));

        $this->assertEmpty($http->requests);
    }

    // ─────────────────────────────────────────────────────────────────
    // getStats
    // ─────────────────────────────────────────────────────────────────

    public function testGetStatsComputesConversationsResponseRateAndCsat(): void
    {
        $pp = new PocketPing();

        $a = $pp->handleConnect(new ConnectRequest(visitorId: 'va'))->sessionId;
        $b = $pp->handleConnect(new ConnectRequest(visitorId: 'vb'))->sessionId;

        // Session A: visitor message + operator reply + 5-star rating.
        $pp->handleMessage(new SendMessageRequest(sessionId: $a, content: 'hi', sender: Sender::VISITOR));
        $pp->sendOperatorMessage($a, 'hello!');
        $pp->handleCsat(new CsatRequest(sessionId: $a, score: 5));

        // Session B: visitor message only (unanswered).
        $pp->handleMessage(new SendMessageRequest(sessionId: $b, content: 'anyone?', sender: Sender::VISITOR));

        $stats = $pp->getStats();

        $this->assertSame(2, $stats['conversations']);
        $this->assertSame(0.5, $stats['responseRate']);
        $this->assertSame(1, $stats['unansweredNow']);
        $this->assertEquals(['percent' => 1, 'average' => 5, 'responses' => 1], $stats['csat']);
        $this->assertCount(7, $stats['conversationsSparkline']);
    }

    public function testGetStatsThrowsWhenStorageCannotListSessions(): void
    {
        $storage = new class implements \PocketPing\Storage\StorageInterface {
            public function createSession(Session $session): void
            {
            }

            public function getSession(string $sessionId): ?Session
            {
                return null;
            }

            public function updateSession(Session $session): void
            {
            }

            public function deleteSession(string $sessionId): void
            {
            }

            public function saveMessage(\PocketPing\Models\Message $message): void
            {
            }

            public function getMessages(string $sessionId, ?string $after = null, int $limit = 50): array
            {
                return [];
            }

            public function getMessage(string $messageId): ?\PocketPing\Models\Message
            {
                return null;
            }

            public function cleanupOldSessions(\DateTimeInterface $olderThan): int
            {
                return 0;
            }

            public function getSessionByVisitorId(string $visitorId): ?Session
            {
                return null;
            }
        };

        $pp = new PocketPing(storage: $storage);

        $this->expectException(\RuntimeException::class);
        $this->expectExceptionMessageMatches('/listSessions/');
        $pp->getStats();
    }

    public function testGetStatsRespectsExplicitWindow(): void
    {
        $storage = new MemoryStorage();
        $pp = new PocketPing(storage: $storage);

        // A session created 10 days ago should fall outside a "last 7 days" default window.
        $old = new Session(
            id: 'old',
            visitorId: 'vo',
            createdAt: new \DateTimeImmutable('-10 days'),
            lastActivity: new \DateTimeImmutable('-10 days'),
        );
        $storage->createSession($old);

        $recent = $pp->handleConnect(new ConnectRequest(visitorId: 'vr'))->sessionId;
        $this->assertNotEmpty($recent);

        $stats = $pp->getStats();
        $this->assertSame(1, $stats['conversations']); // only the recent one
    }
}
